controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1199 @@
+			for {
+				if err := connect(); err == nil {
+					continue outer
+				}
+				time.Sleep(100 * time.Millisecond)
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *Scheduler) Run() error {
+	log := fnLogger()
+	log.Info("starting scheduler loop")
+	defer log.Info("scheduler loop exited")
+
+	if err := s.streamHostEvents(); err != nil {
+		return err
+	}
+
+	if err := s.streamFormationEvents(); err != nil {
+		return err
+	}
+
+	s.tickSyncJobs(30 * time.Second)
+	s.tickSyncFormations(time.Minute)
+
+	go s.RunPutJobs()
+
+	for {
+		select {
+		case <-s.stop:
+			log.Info("stopping scheduler loop")
+			close(s.putJobs)
+			return nil
+		case isLeader := <-s.changeLeader:
+			s.HandleLeaderChange(isLeader)
+			continue
+		default:
+		}
+
+		// Handle events that reconcile scheduler state with the cluster
+		select {
+		case req := <-s.jobRequests:
+			s.HandleJobRequest(req)
+			continue
+		case e := <-s.hostEvents:
+			s.HandleHostEvent(e)
+			continue
+		case e, ok := <-s.jobEvents:
+			if !ok {
+				return errors.New("job events channel closed prematurely")
+			}
+			s.HandleJobEvent(e)
+			continue
+		case f := <-s.formationEvents:
+			s.HandleFormationChange(f)
+			continue
+		default:
+		}
+
+		// Handle sync events
+		select {
+		case <-s.syncFormations:
+			s.SyncFormations()
+			continue
+		case <-s.syncJobs:
+			s.SyncJobs()
+			continue
+		default:
+		}
+
+		// Finally, handle triggering cluster changes.
+		// Re-select on all the channels so we don't have to sleep nor spin
+		select {
+		case <-s.rectify:
+			if !s.ready {
+				continue
+			}
+			s.Rectify()
+		case <-s.stop:
+			log.Info("stopping scheduler loop")
+			close(s.putJobs)
+			return nil
+		case isLeader := <-s.changeLeader:
+			s.HandleLeaderChange(isLeader)
+		case req := <-s.jobRequests:
+			s.HandleJobRequest(req)
+		case e := <-s.hostEvents:
+			s.HandleHostEvent(e)
+		case e, ok := <-s.jobEvents:
+			if !ok {
+				return errors.New("job events channel closed prematurely")
+			}
+			s.HandleJobEvent(e)
+		case f := <-s.formationEvents:
+			s.HandleFormationChange(f)
+		case <-s.syncFormations:
+			s.SyncFormations()
+		case <-s.syncJobs:
+			s.SyncJobs()
+		}
+	}
+	return nil
+}
+
+func (s *Scheduler) SyncJobs() {
+	defer s.sendEvent(NewEvent(EventTypeClusterSync, nil, nil))
+
+	log := fnLogger()
+	log.Info("syncing jobs")
+
+	log.Info("getting host list")
+	hosts, err := s.getHosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return
+	}
+
+	newJobs := make(Jobs)
+	for _, h := range hosts {
+		hostLog := log.New("host.id", h.ID())
+
+		hostLog.Info(fmt.Sprintf("getting jobs for host %s", h.ID()))
+		activeJobs, err := h.ListJobs()
+		if err != nil {
+			hostLog.Error("error getting jobs list", "err", err)
+			continue
+		}
+		hostLog.Info(fmt.Sprintf("got %d active job(s) for host %s", len(activeJobs), h.ID()))
+
+		for _, job := range activeJobs {
+			s.handleActiveJob(&job)
```

Comment:
Ok, but this still needs fixing right? I think `SyncJobs` should be explicit in what it is doing (i.e. just making sure that `s.jobs` and the controller are in sync with the cluster), so I feel like things would be clearer if this were something like:

```go
for _, job := range jobs {
  if j, ok := s.jobs[job.ID]; ok && j.State != job.State {
    // the state has changed from what we know, treat it like an event
    s.HandleJobEvent(job)
  }
}
```

I've seen hosts with thousands of active jobs, so potentially doing thousands of unnecessary `PutJob` calls every 30s feels wrong.